{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "KEGG\n",
        "====\n",
        "\n",
        "KEGG (<http://www.kegg.jp/>) is a database resource for understanding\n",
        "high-level functions and utilities of the biological system, such as the\n",
        "cell, the organism and the ecosystem, from molecular-level information,\n",
        "especially large-scale molecular datasets generated by genome sequencing\n",
        "and other high-throughput experimental technologies.\n",
        "\n",
        "Please note that the KEGG parser implementation in Biopython is\n",
        "incomplete. While the KEGG website indicates many flat file formats,\n",
        "only parsers and writers for compound, enzyme, and map are currently\n",
        "implemented. However, a generic parser is implemented to handle the\n",
        "other formats.\n",
        "\n",
        "Parsing KEGG records\n",
        "--------------------\n",
        "\n",
        "Parsing a KEGG record is as simple as using any other file format parser\n",
        "in Biopython. (Before running the following codes, please open\n",
        "http://rest.kegg.jp/get/ec:5.4.2.2 with your web browser and save it as\n",
        "ec\\_5.4.2.2.txt.)"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "KEGG\n",
        "\n",
        " ====\n",
        "\n",
        " KEGG ( [http://www.kegg.jp/](http://www.kegg.jp/) ) é um recurso de banco de dados para compreensão\n",
        "\n",
        " funções e utilidades de alto nível do sistema biológico, como o\n",
        "\n",
        " célula, o organismo e o ecossistema, a partir de informações em nível molecular,\n",
        "\n",
        " especialmente conjuntos de dados moleculares em grande escala gerados por sequenciamento do genoma\n",
        "\n",
        " e outras tecnologias experimentais de alto rendimento.\n",
        "\n",
        " Observe que a implementação do analisador KEGG no Biopython é\n",
        "\n",
        " incompleto. Embora o site KEGG indique muitos formatos de arquivo simples,\n",
        "\n",
        " apenas analisadores e gravadores para compostos, enzimas e mapas são atualmente\n",
        "\n",
        " implementado. No entanto, um analisador genérico é implementado para lidar com o\n",
        "\n",
        " outros formatos.\n",
        "\n",
        " Análise de registros KEGG\n",
        "\n",
        " Analisar um registro KEGG é tão simples quanto usar qualquer outro analisador de formato de arquivo\n",
        "\n",
        " em Biopython. (Antes de executar os códigos a seguir, abra\n",
        "\n",
        " [http://rest.kegg.jp/get/ec:5.4.2.2](http://rest.kegg.jp/get/ec:5.4.2.2) com seu navegador e salve-o como\n",
        "\n",
        " ec_5.4.2.2.txt.)\n"
      ],
      "metadata": {}
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "--2016-01-12 20:44:45--  http://rest.kegg.jp/get/ec:5.4.2.2\n",
            "Resolving rest.kegg.jp (rest.kegg.jp)... 133.103.200.77\n",
            "Connecting to rest.kegg.jp (rest.kegg.jp)|133.103.200.77|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: unspecified [text/plain]\n",
            "Saving to: ‘ec_5.4.2.2.txt’\n",
            "\n",
            "ec_5.4.2.2.txt          [   <=>                ] 105.99K   141KB/s   in 0.8s   \n",
            "\n",
            "2016-01-12 20:44:46 (141 KB/s) - ‘ec_5.4.2.2.txt’ saved [108534]\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!wget http://rest.kegg.jp/get/ec:5.4.2.2 -O ec_5.4.2.2.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "['Isomerases;',\n",
              " 'Intramolecular transferases;',\n",
              " 'Phosphotransferases (phosphomutases)']"
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "from Bio.KEGG import Enzyme\n",
        "\n",
        "records = Enzyme.parse(open(\"ec_5.4.2.2.txt\"))\n",
        "\n",
        "record = list(records)[0]\n",
        "\n",
        "record.classname"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'5.4.2.2'"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "record.entry"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "The following section will shows how to download the above enzyme using\n",
        "the KEGG api as well as how to use the generic parser with data that\n",
        "does not have a custom parser implemented.\n",
        "\n",
        "Querying the KEGG API\n",
        "---------------------\n",
        "\n",
        "Biopython has full support for the querying of the KEGG api. Querying\n",
        "all KEGG endpoints are supported; all methods documented by KEGG\n",
        "(<http://www.kegg.jp/kegg/rest/keggapi.html>) are supported. The\n",
        "interface has some validation of queries which follow rules defined on\n",
        "the KEGG site. However, invalid queries which return a 400 or 404 must\n",
        "be handled by the user.\n",
        "\n",
        "First, here is how to extend the above example by downloading the\n",
        "relevant enzyme and passing it through the Enzyme parser."
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "A seção a seguir mostra como baixar a enzima acima usando\n",
        "\n",
        " a API KEGG, bem como como usar o analisador genérico com dados que\n",
        "\n",
        " não tem um analisador customizado implementado.\n",
        "\n",
        " Consultando a API KEGG\n",
        "\n",
        " Biopython tem suporte total para a consulta da API KEGG. Consultando\n",
        "\n",
        " todos os endpoints KEGG são suportados; todos os métodos documentados por KEGG\n",
        "\n",
        " ( [http://www.kegg.jp/kegg/rest/keggapi.html](http://www.kegg.jp/kegg/rest/keggapi.html) ) são suportados. o\n",
        "\n",
        " interface tem alguma validação de consultas que seguem regras definidas em\n",
        "\n",
        " o site KEGG. No entanto, consultas inválidas que retornam 400 ou 404 devem\n",
        "\n",
        " ser manipulado pelo usuário.\n",
        "\n",
        " Primeiro, aqui está como estender o exemplo acima baixando o\n",
        "\n",
        " enzima relevante e passando-a através do analisador de enzimas.\n"
      ],
      "metadata": {}
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "108534"
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "from Bio.KEGG import REST\n",
        "\n",
        "from Bio.KEGG import Enzyme\n",
        "\n",
        "request = REST.kegg_get(\"ec:5.4.2.2\")\n",
        "\n",
        "open(\"ec_5.4.2.2.txt\", 'w').write(request.read().decode(\"utf-8\"))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "['Isomerases;',\n",
              " 'Intramolecular transferases;',\n",
              " 'Phosphotransferases (phosphomutases)']"
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "records = Enzyme.parse(open(\"ec_5.4.2.2.txt\"))\n",
        "\n",
        "record = list(records)[0]\n",
        "\n",
        "record.classname"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'5.4.2.2'"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "record.entry"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Now, here’s a more realistic example which shows a combination of\n",
        "querying the KEGG API. This will demonstrate how to extract a unique set\n",
        "of all human pathway gene symbols which relate to DNA repair. The steps\n",
        "that need to be taken to do so are as follows. First, we need to get a\n",
        "list of all human pathways. Secondly, we need to filter those for ones\n",
        "which relate to “repair”. Lastly, we need to get a list of all the gene\n",
        "symbols in all repair pathways."
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "Agora, aqui está um exemplo mais realista que mostra uma combinação de\n",
        "\n",
        " consultar a API KEGG. Isso irá demonstrar como extrair um conjunto único\n",
        "\n",
        " de todos os símbolos de genes da via humana que se relacionam com a reparação do DNA Os passos\n",
        "\n",
        " que precisam ser considerados para fazê-lo são os seguintes. Primeiro, precisamos obter um\n",
        "\n",
        " lista de todas as vias humanas. Em segundo lugar, precisamos filtrar aqueles para aqueles\n",
        "\n",
        " que se relacionam com “reparo”. Por último, precisamos obter uma lista de todos os genes\n",
        "\n",
        " símbolos em todas as vias de reparo.\n"
      ],
      "metadata": {}
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "['path:hsa00010\\tGlycolysis / Gluconeogenesis - Homo sapiens (human)',\n",
              " 'path:hsa00020\\tCitrate cycle (TCA cycle) - Homo sapiens (human)',\n",
              " 'path:hsa00030\\tPentose phosphate pathway - Homo sapiens (human)',\n",
              " 'path:hsa00040\\tPentose and glucuronate interconversions - Homo sapiens (human)',\n",
              " 'path:hsa00051\\tFructose and mannose metabolism - Homo sapiens (human)']"
            ]
          },
          "execution_count": 8,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "from Bio.KEGG import REST\n",
        "\n",
        "human_pathways = REST.kegg_list(\"pathway\", \"hsa\").read()\n",
        "\n",
        "human_pathways.decode(\"utf-8\").split(\"\\n\")[0:5]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "['path:hsa03410', 'path:hsa03420', 'path:hsa03430']"
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter all human pathways for repair pathways\n",
        "repair_pathways = []\n",
        "for line in human_pathways.decode(\"utf-8\").rstrip().split(\"\\n\"):\n",
        "    entry, description = line.split(\"\\t\")\n",
        "    if \"repair\" in description:\n",
        "        repair_pathways.append(entry)\n",
        "\n",
        "repair_pathways"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "There are 3 repair pathways and 78 repair genes. The genes are:\n",
            "OGG1, NTHL1, NEIL1, NEIL2, NEIL3, UNG, TDG, SMUG1, MUTYH, MPG, MBD4, APEX1, APEX2, POLB, POLL, HMGB1, XRCC1, PCNA, POLD1, POLD2, POLD3, POLD4, POLE, POLE2, POLE3, POLE4, LIG1, LIG3, PARP2, PARP1, PARP3, PARP4, FEN1, RBX1, CUL4B, CUL4A, DDB1, DDB2, XPC, RAD23B, RAD23A, CETN2, ERCC8, ERCC6, CDK7, MNAT1, CCNH, ERCC3, ERCC2, GTF2H5, GTF2H1, GTF2H2, GTF2H2C_2, GTF2H2C, GTF2H3, GTF2H4, ERCC5, BIVM-ERCC5, XPA, RPA1, RPA2, RPA3, RPA4, ERCC4, ERCC1, RFC1, RFC4, RFC2, RFC5, RFC3, SSBP1, PMS2, MLH1, MSH6, MSH2, MSH3, MLH3, EXO1\n"
          ]
        }
      ],
      "source": [
        "# Get the genes for pathways and add them to a list\n",
        "repair_genes = [] \n",
        "for pathway in repair_pathways:\n",
        "    pathway_file = REST.kegg_get(pathway).read()  # query and read each pathway\n",
        "\n",
        "    # iterate through each KEGG pathway file, keeping track of which section\n",
        "    # of the file we're in, only read the gene in each pathway\n",
        "    current_section = None\n",
        "    for line in pathway_file.decode(\"utf-8\").rstrip().split(\"\\n\"):\n",
        "        section = line[:12].strip()  # section names are within 12 columns\n",
        "        if not section == \"\":\n",
        "            current_section = section\n",
        "        \n",
        "        if current_section == \"GENE\":\n",
        "            gene_identifiers, gene_description = line[12:].split(\"; \")\n",
        "            gene_id, gene_symbol = gene_identifiers.split()\n",
        "\n",
        "            if not gene_symbol in repair_genes:\n",
        "                repair_genes.append(gene_symbol)\n",
        "\n",
        "print(\"There are %d repair pathways and %d repair genes. The genes are:\" % \\\n",
        "        (len(repair_pathways), len(repair_genes)))\n",
        "print(\", \".join(repair_genes))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "The KEGG API wrapper is compatible with all endpoints. Usage is\n",
        "essentially replacing all slashes in the url with commas and using that\n",
        "list as arguments to the corresponding method in the KEGG module. Here\n",
        "are a few examples from the api documentation\n",
        "(<http://www.kegg.jp/kegg/docs/keggapi.html>).\n",
        "\n",
        "    /list/hsa:10458+ece:Z5100            -> REST.kegg_list([\"hsa:10458\", \"ece:Z5100\"])\n",
        "    /find/compound/300-310/mol_weight    -> REST.kegg_find(\"compound\", \"300-310\", \"mol_weight\")\n",
        "    /get/hsa:10458+ece:Z5100/aaseq      -> REST.kegg_get([\"hsa:10458\", \"ece:Z5100\"], \"aaseq\")"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "O wrapper KEGG API é compatível com todos os terminais. O uso é\n",
        "\n",
        " essencialmente substituindo todas as barras no url por vírgulas e usando\n",
        "\n",
        " lista como argumentos para o método correspondente no módulo KEGG. Aqui\n",
        "\n",
        " são alguns exemplos da documentação da API\n",
        "\n",
        " ( [http://www.kegg.jp/kegg/docs/keggapi.html](http://www.kegg.jp/kegg/docs/keggapi.html) ).\n"
      ],
      "metadata": {}
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {},
  "nbformat": 4,
  "nbformat_minor": 0
}